### DISCUSSION TITLE
Use C# 8.0 features?

### DISCUSSION THREAD

**Discussion by Numpsy** [icsharpcode/SharpZipLib#632] [Ideas]
One of the things that #501 did was to set the C# language version to 8.0 (to get the nullability features)

I was just wondering if there was any general thought about turning that on seperately?

(Just a thought I had when looking at some analyzer warnings and thinking 'the simplified ```using``` syntax might be nice in some places')

**Comment by Numpsy** [comment]
e.g. changing the indentation in 

https://github.com/Numpsy/SharpZipLib/commit/23172d849891dd0b7ee5b86fc1ccaa095feabb0b

makes the diff viewer  show many differences, when it's really just some ```using```s with braces changing the indents

**Reply by piksel** [reply]
It only works in PRs, but you can actually hide those:
![image](https://user-images.githubusercontent.com/807383/117961791-cc206d00-b31e-11eb-9fa5-5ab277a2a6f8.png)

**Comment by piksel** [comment]
I didn't think compiling for .NET 4.5 would work using C# 8, but it seems like it does... It could definitely be done separately, but that would cause many warnings to appear (or, maybe that's just resharper?). But perhaps that is fine, and solving those warnings could be done in subsequent PRs?

**Reply by piksel** [reply]
This answer explains it really well (and thoroughly!):
https://stackoverflow.com/a/57020770/910069

In short: Visual Studio doesn't let you switch the language version since all features are not supported, but most (including the nullable reference types, which was my main concern) does work.

**Reply by christophwille** [reply]
Maybe you should start the discussion like we did (ILSpy, AvalonEdit) to cut support for old frameworks. Eg. 472 will the lowest supported version from MS next year (we always tend to keep in line with min. officially supported by MS)

**Reply by Numpsy** [reply]
fwiw, I've been working on a product that builds against .NET 4.7.2, and that uses nullability annotations, 'simplified usings', async enumerable and some other minor bits with no issues that i've seen. (and you can avoid using the bits that need extra packages and just keep it to the syntax sugar ones).

>(or, maybe that's just resharper?)

It tends to result in Visual Studio making suggestions (rather than warnings) to simplify all usings, or to replace things with switch expressions etc (similar to how it currently suggests making many fields readonly) , though I think you can control all that via .editorconfig.

**Reply by piksel** [reply]
Yeah, I meant nullability explicitly, just setting the version to 8 shouldn't cause any warnings, but using some features might 😄

I don't think there should be that much of a difference between .NET 4.5 and 4.7 from a C# 8 view, but yeah, we should re-justify supporting .NET 4.5 in upcoming versions...

The only things to look out for afaik. are compile- (which will prevent merging PRs anyway) and performance issues. Basically, all the C# 8 sugar *work* on .NET FW, but have not been optimized for.
